### PR TITLE
invalidate session in rp when a bcl request is received

### PR DESCRIPTION
**Pull request by jimmy1wu**
for #20360 

creates a new oidc session cache which caches oidc sessions based on the sub, sid, and oidc session id. the sub and sid are from the id token returned from the op, while the oidc session id is something we generate to keep track of the sessions on our rp. oidc session id's without a sid will be grouped together for that sub.

e.g., `sub1` has `oidcsessionid1` and `oidcsessionid2` which do not have a sid, but also has `oidcsessionid3` which belongs to `sid1` and `oidcsessionid4` which belongs to `sid2`.
```
{
    "sub1": {
        null: ["oidcsessionid1", "oidcsessionid2"],
        "sid1": ["oidcsessionid3"],
        "sid2": ["oidcsessionid4"]
    }
}
```

oidc session id's will hold information about the session in the format: `configId:sub:sid:timestamp` and are stored in the oidc session cache, as well as in a `WASOidcSession` cookie.

additionally, a set of all invalidated oidc session id's is also maintained in the cache to allow for constant time searches to see if the oidc session id has been invalidated. invalidated sessions need to be fully logged out.

**insert scenerio:** 
after a successful oidc flow, the oidc session id is created and stored in the oidc session cache using the sub claim and sid claim (if it's available) from the id token. the oidc session id will also be set in the user agent's cookies under the cookie name `WASOidcSession` to be used later.

**invalidate scenerio:** 
after the rp receives a back-channel logout request, the oidc sessions are removed based on the `sub` and `sid` claims in the logout token. if a `sid` claim is available, only that particular oidc session is invalidated. if a `sid` claim is _not_ available, all the oidc sessions belonging to that `sub` are invalidated. sessions are invalidated by removing the oidc session id's from the map and then storing them in the invalidated sessions set.

**search and delete scenerio:** 
every request to `WebAuthenticatorProxy` will check if the session (stored in the `WASOidcSession` cookie) has been invalidated when it gets to `handleOidcClient`. if the session has been invalidated, then a full logout is performed and then the oidc session id is removed from the invalidated session set. if the session has not been invalidated, nothing happens here.

--- 
**TODOS**

- the oidc session id will need to be encrypted using the client secret to obfuscate the information it holds

- need to invalidate the oidc sessions when a non-backchannel logout happens as well.

- we only currently implement an `InMemoryOidcSessionCache` and we need to implement a j-cache version of `OidcSessionCache`, so this can work in a distributed scenerio as well.

### CODE REVIEW

File: dev/com.ibm.ws.security.openidconnect.client/src/com/ibm/ws/security/openidconnect/client/internal/OidcClientImpl.java [hunk 1]
```diff
@@ -430,6 +432,28 @@ public ProviderAuthenticationResult authenticate(HttpServletRequest req,
...
 
+    @Override
+    public void logoutIfSessionInactive(HttpServletRequest req, String provider) {
+        OidcClientConfig oidcClientConfig = oidcClientConfigRef.getService(provider);
+        if (!oidcClientConfig.isBackchannelLogoutSupported()) {
+            return;
+        }
+
+        HttpSessionCache httpSessionCache = oidcClientConfig.getHttpSessionCache();
+        String httpSessionId = req.getSession().getId();
```
Comment (ayoho) [inline dev/com.ibm.ws.security.openidconnect.client/src/com/ibm/ws/security/openidconnect/client/internal/OidcClientImpl.java]: This might need to be something like:
```suggestion
        HttpSession httpSession = req.getSession(false);
        String httpSessionId = (httpSession == null) ? null : httpSession.getId();
```

File: dev/com.ibm.ws.security.openidconnect.clients.common/src/com/ibm/ws/security/openidconnect/backchannellogout/BackchannelLogoutHelper.java [hunk 2]
```diff
@@ -77,13 +79,21 @@ String validateRequestAndGetLogoutTokenParameter() throws BackchannelLogoutExcep
...
-        // TODO
+    void performLogout(JwtClaims logoutTokenClaims) throws BackchannelLogoutException {
+        try {
+            String sub = logoutTokenClaims.getSubject();
+            String sid = logoutTokenClaims.getClaimValue("sid", String.class);
+
+            HttpSessionCache httpSessionCache = clientConfig.getOidcClientConfig().getHttpSessionCache();
+            httpSessionCache.removeSession(sub, sid);
+        } catch (Exception e) {
+            throw new BackchannelLogoutException(e.getMessage(), HttpServletResponse.SC_NOT_IMPLEMENTED);
```
Comment (ayoho) [inline dev/com.ibm.ws.security.openidconnect.clients.common/src/com/ibm/ws/security/openidconnect/backchannellogout/BackchannelLogoutHelper.java]: This could probably use its own NLS message saying something like "The OpenID Connect client encountered the following error while performing back-channel logout for user {0}: {1}"

File: dev/com.ibm.ws.security.openidconnect.clients.common/src/com/ibm/ws/security/openidconnect/client/jose4j/util/Jose4jUtil.java [hunk 3]
```diff
@@ -183,6 +188,18 @@ public ProviderAuthenticationResult createResultWithJose4J(String responseState,
         return oidcResult;
     }
 
+    private void addHttpSessionToCache(OidcClientRequest oidcClientRequest, JwtClaims jwtClaims, ConvergedClientConfig clientConfig) throws MalformedClaimException {
+        OidcClientConfig oidcClientConfig = clientConfig.getOidcClientConfig();
+        if (oidcClientConfig.isBackchannelLogoutSupported()) {
+            String sub = jwtClaims.getSubject();
+            String sid = jwtClaims.getClaimValue("sid", String.class);
+            String httpSessionId = oidcClientRequest.getRequest().getSession().getId();
```
Comment (ayoho) [inline dev/com.ibm.ws.security.openidconnect.clients.common/src/com/ibm/ws/security/openidconnect/client/jose4j/util/Jose4jUtil.java]: Similar comment here as before. Not sure if we need to do something like this:
```suggestion
            HttpSession session = oidcClientRequest.getRequest().getSession(false);
            String httpSessionId = (session == null) ? null : session.getId();
```

File: dev/com.ibm.ws.webcontainer.security/test/com/ibm/ws/webcontainer/security/InMemoryHttpSessionCacheTest.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ *
+ */
+public class InMemoryHttpSessionCacheTest {
```
Comment (ayoho) [inline dev/com.ibm.ws.webcontainer.security/test/com/ibm/ws/webcontainer/security/InMemoryHttpSessionCacheTest.java]: Depending on what we decide as far as using `request.getSession(false)` goes, we might need to add unit tests to verify behavior for when an HTTP session ID is null.

File: dev/com.ibm.ws.security.openidconnect.clients.common/src/com/ibm/ws/security/openidconnect/backchannellogout/BackchannelLogoutHelper.java (hunk 2 above)
Comment (jimmy1wu) [reply dev/com.ibm.ws.security.openidconnect.clients.common/src/com/ibm/ws/security/openidconnect/backchannellogout/BackchannelLogoutHelper.java]: i can't include the user/sub in the message since retrieving the claim value may throw an error. included the entire logout token claims set instead.

File: dev/com.ibm.ws.security.openidconnect.client/src/com/ibm/ws/security/openidconnect/client/internal/OidcClientImpl.java (hunk 1 above)
Comment (jimmy1wu) [reply dev/com.ibm.ws.security.openidconnect.client/src/com/ibm/ws/security/openidconnect/client/internal/OidcClientImpl.java]: changed it so it doesn't try to create a session if one doesn't already exist.

File: dev/com.ibm.ws.security.openidconnect.clients.common/src/com/ibm/ws/security/openidconnect/client/jose4j/util/Jose4jUtil.java (hunk 3 above)
Comment (jimmy1wu) [reply dev/com.ibm.ws.security.openidconnect.clients.common/src/com/ibm/ws/security/openidconnect/client/jose4j/util/Jose4jUtil.java]: ditto

File: dev/com.ibm.ws.webcontainer.security/test/com/ibm/ws/webcontainer/security/InMemoryHttpSessionCacheTest.java (hunk 4 above)
Comment (jimmy1wu) [reply dev/com.ibm.ws.webcontainer.security/test/com/ibm/ws/webcontainer/security/InMemoryHttpSessionCacheTest.java]: i added a check to see if an httpsessionid is null/empty for now before inserting. still need to investigate a way to use some other sort of id when an http session is not available.

**Comment by jimmy1wu** [comment]
#libby

**Comment by jimmy1wu** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.security.oidc.client_fat.1,com.ibm.ws.security.oidc.client_fat.2,com.ibm.ws.security.oidc.client_fat.config,com.ibm.ws.security.oidc.client_fat.e2e,com.ibm.ws.security.oidc.client_fat.jaxrs,com.ibm.ws.security.oidc.client_fat.saml,com.ibm.ws.security.oidc.client_fat.spnego,com.ibm.ws.security.oidc.server_fat,com.ibm.ws.security.oidc.server_fat.app-passwords,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.clientregistration,com.ibm.ws.security.oidc.server_fat.config,com.ibm.ws.security.oidc.server_fat.config.oauth,com.ibm.ws.security.oidc.server_fat.config.oidc,com.ibm.ws.security.oidc.server_fat.endpoint.authorize,com.ibm.ws.security.oidc.server_fat.endpoint.coveragemap,com.ibm.ws.security.oidc.server_fat.endpoint.discovery,com.ibm.ws.security.oidc.server_fat.endpoint.end_session,com.ibm.ws.security.oidc.server_fat.endpoint.introspect,com.ibm.ws.security.oidc.server_fat.endpoint.revoke,com.ibm.ws.security.oidc.server_fat.endpoint.token,com.ibm.ws.security.oidc.server_fat.endpoint.userinfo,com.ibm.ws.security.oidc.server_fat.feature,com.ibm.ws.security.oidc.server_fat.granttype.jwt,com.ibm.ws.security.oidc.server_fat.gui.tokenMgmt,com.ibm.ws.security.oidc.server_fat.jaxrs,com.ibm.ws.security.oidc.server_fat.jaxrs.config,com.ibm.ws.security.oidc.server_fat.oauth,com.ibm.ws.security.oidc.server_fat.oidc,com.ibm.ws.security.oidc.server_fat.saml,com.ibm.ws.security.social_fat,com.ibm.ws.security.social_fat.LibertyOP.1,com.ibm.ws.security.social_fat.LibertyOP.2,com.ibm.ws.security.social_fat.OpenShift,com.ibm.ws.security.social_fat.delegated,com.ibm.ws.security.social_fat.multiProvider

**Review by ayoho — APPROVED** [review]

File: dev/com.ibm.ws.security.openidconnect.clients.common/resources/com/ibm/ws/security/openidconnect/clients/common/resources/OidcClientMessages.nlsprops [hunk 5]
```diff
@@ -297,5 +297,9 @@ LOGOUT_TOKEN_SID_REQUIRED_BUT_MISSING=CWWKS1551E: The logout token is missing th
 LOGOUT_TOKEN_SID_REQUIRED_BUT_MISSING.explanation=The OpenID Connect client requires that the logout token contains a session ID to identify the OpenID Connect client session with the OpenID Connect provider.
 LOGOUT_TOKEN_SID_REQUIRED_BUT_MISSING.useraction=Set the backchannelLogoutSessionRequired OpenID Connect client configuration attribute to false, or ensure that the logout token contains a "sid" claim.
 
+BACKCHANNEL_LOGOUT_PERFORM_LOGOUT_FAILED=CWWKS1552E: The OpenID Connect client encountered the following error while performing back-channel logout with a logout token with claims {0}: {1}
```
Comment (helyarp) [inline dev/com.ibm.ws.security.openidconnect.clients.common/resources/com/ibm/ws/security/openidconnect/clients/common/resources/OidcClientMessages.nlsprops]: Change "the following" to "an" 

(nit, not a required change)

File: dev/com.ibm.ws.security.openidconnect.clients.common/resources/com/ibm/ws/security/openidconnect/clients/common/resources/OidcClientMessages.nlsprops [hunk 6]
```diff
@@ -297,5 +297,9 @@ LOGOUT_TOKEN_SID_REQUIRED_BUT_MISSING=CWWKS1551E: The logout token is missing th
...
+BACKCHANNEL_LOGOUT_PERFORM_LOGOUT_FAILED.explanation=The back-channel logout request did not complete successfully.
+BACKCHANNEL_LOGOUT_PERFORM_LOGOUT_FAILED.useraction=View the error message for more information.
```
Comment (helyarp) [inline dev/com.ibm.ws.security.openidconnect.clients.common/resources/com/ibm/ws/security/openidconnect/clients/common/resources/OidcClientMessages.nlsprops]: Can you provide more specifics about what the user needs to do? What will the error message say?

**Review by helyarp — COMMENTED** [review]
Suggested change and more info.

File: dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/WebProviderAuthenticatorProxy.java [hunk 7]
```diff
@@ -587,18 +587,20 @@ private AuthenticationResult handleOidcClient(HttpServletRequest req, HttpServle
         if (oidcClient == null) {
             return new AuthenticationResult(AuthResult.CONTINUE, "OpenID Connect client is not available, skipping OpenID Connect client...");
         }
+        String provider = oidcClient.getOidcProvider(req);
+        if (provider == null) {
+            return new AuthenticationResult(AuthResult.CONTINUE, "not an OpenID Connect client request, skipping OpenID Connect client...");
+        }
 
         if (firstCall) {
+            oidcClient.logoutIfSessionInactive(req, provider);
```
Comment (teddyjtorres) [inline dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/WebProviderAuthenticatorProxy.java]: Please add a beta guard. Thanks!

Comment (jimmy1wu) [reply dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/WebProviderAuthenticatorProxy.java]: beta guard added to `Jose4jUtil` and `OidcClientImpl`

File: dev/com.ibm.ws.security.openidconnect.clients.common/resources/com/ibm/ws/security/openidconnect/clients/common/resources/OidcClientMessages.nlsprops (hunk 5 above)
Comment (jimmy1wu) [reply dev/com.ibm.ws.security.openidconnect.clients.common/resources/com/ibm/ws/security/openidconnect/clients/common/resources/OidcClientMessages.nlsprops]: the nls is no longer needed.
i've removed it from the file.

File: dev/com.ibm.ws.security.openidconnect.clients.common/resources/com/ibm/ws/security/openidconnect/clients/common/resources/OidcClientMessages.nlsprops (hunk 6 above)
Comment (jimmy1wu) [reply dev/com.ibm.ws.security.openidconnect.clients.common/resources/com/ibm/ws/security/openidconnect/clients/common/resources/OidcClientMessages.nlsprops]: the nls is no longer needed.
i've removed it from the file.

**Comment by jimmy1wu** [comment]
#libby

**Comment by jimmy1wu** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.security.oidc.client_fat.1,com.ibm.ws.security.oidc.client_fat.2,com.ibm.ws.security.oidc.client_fat.config,com.ibm.ws.security.oidc.client_fat.e2e,com.ibm.ws.security.oidc.client_fat.jaxrs,com.ibm.ws.security.oidc.client_fat.saml,com.ibm.ws.security.oidc.client_fat.spnego,com.ibm.ws.security.oidc.server_fat,com.ibm.ws.security.oidc.server_fat.app-passwords,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.clientregistration,com.ibm.ws.security.oidc.server_fat.config,com.ibm.ws.security.oidc.server_fat.config.oauth,com.ibm.ws.security.oidc.server_fat.config.oidc,com.ibm.ws.security.oidc.server_fat.endpoint.authorize,com.ibm.ws.security.oidc.server_fat.endpoint.coveragemap,com.ibm.ws.security.oidc.server_fat.endpoint.discovery,com.ibm.ws.security.oidc.server_fat.endpoint.end_session,com.ibm.ws.security.oidc.server_fat.endpoint.introspect,com.ibm.ws.security.oidc.server_fat.endpoint.revoke,com.ibm.ws.security.oidc.server_fat.endpoint.token,com.ibm.ws.security.oidc.server_fat.endpoint.userinfo,com.ibm.ws.security.oidc.server_fat.feature,com.ibm.ws.security.oidc.server_fat.granttype.jwt,com.ibm.ws.security.oidc.server_fat.gui.tokenMgmt,com.ibm.ws.security.oidc.server_fat.jaxrs,com.ibm.ws.security.oidc.server_fat.jaxrs.config,com.ibm.ws.security.oidc.server_fat.oauth,com.ibm.ws.security.oidc.server_fat.oidc,com.ibm.ws.security.oidc.server_fat.saml,com.ibm.ws.security.social_fat,com.ibm.ws.security.social_fat.LibertyOP.1,com.ibm.ws.security.social_fat.LibertyOP.2,com.ibm.ws.security.social_fat.OpenShift,com.ibm.ws.security.social_fat.delegated,com.ibm.ws.security.social_fat.multiProvider

File: dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/InMemoryOidcSessionCache.java [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+        }
+
+        public boolean insertSession(String sid, String oidcSessionId) {
+            if (!sidToSessionsMap.containsKey(sid)) {
+                sidToSessionsMap.put(sid, new ArrayList<>());
+            }
+            List<String> sessions = sidToSessionsMap.get(sid);
+            if (sid != null && !sid.isEmpty() && sessions.size() > 0) {
+                return false;
+            }
```
Comment (ayoho) [inline dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/InMemoryOidcSessionCache.java]: So there will only ever be one session associated with a non-null, non-empty `sid`, yeah? The value for a null `sid` will be for all of the sessions that don't have an associated `sid`, which is why the value is a List?

**Review by ayoho — APPROVED** [review]

**Review by ayoho — APPROVED** [review]

File: dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/InMemoryOidcSessionCache.java (hunk 8 above)
Comment (jimmy1wu) [reply dev/com.ibm.ws.webcontainer.security/src/com/ibm/ws/webcontainer/security/InMemoryOidcSessionCache.java]: exactly.
[also included unit tests for these scenerios](https://github.com/OpenLiberty/open-liberty/pull/20879/files#diff-b8e83b75d6d09cd8220bec21488e6e41d80b5e155130a92b6560f15f8d7b41b2R87-R127)

**Review by teddyjtorres — APPROVED** [review]

**Comment by jimmy1wu** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.security.oidc.client_fat.1,com.ibm.ws.security.oidc.client_fat.2,com.ibm.ws.security.oidc.client_fat.config,com.ibm.ws.security.oidc.client_fat.e2e,com.ibm.ws.security.oidc.client_fat.jaxrs,com.ibm.ws.security.oidc.client_fat.saml,com.ibm.ws.security.oidc.client_fat.spnego,com.ibm.ws.security.oidc.server_fat,com.ibm.ws.security.oidc.server_fat.app-passwords,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.clientregistration,com.ibm.ws.security.oidc.server_fat.config,com.ibm.ws.security.oidc.server_fat.config.oauth,com.ibm.ws.security.oidc.server_fat.config.oidc,com.ibm.ws.security.oidc.server_fat.endpoint.authorize,com.ibm.ws.security.oidc.server_fat.endpoint.coveragemap,com.ibm.ws.security.oidc.server_fat.endpoint.discovery,com.ibm.ws.security.oidc.server_fat.endpoint.end_session,com.ibm.ws.security.oidc.server_fat.endpoint.introspect,com.ibm.ws.security.oidc.server_fat.endpoint.revoke,com.ibm.ws.security.oidc.server_fat.endpoint.token,com.ibm.ws.security.oidc.server_fat.endpoint.userinfo,com.ibm.ws.security.oidc.server_fat.feature,com.ibm.ws.security.oidc.server_fat.granttype.jwt,com.ibm.ws.security.oidc.server_fat.gui.tokenMgmt,com.ibm.ws.security.oidc.server_fat.jaxrs,com.ibm.ws.security.oidc.server_fat.jaxrs.config,com.ibm.ws.security.oidc.server_fat.oauth,com.ibm.ws.security.oidc.server_fat.oidc,com.ibm.ws.security.oidc.server_fat.saml,com.ibm.ws.security.social_fat,com.ibm.ws.security.social_fat.LibertyOP.1,com.ibm.ws.security.social_fat.LibertyOP.2,com.ibm.ws.security.social_fat.OpenShift,com.ibm.ws.security.social_fat.delegated,com.ibm.ws.security.social_fat.multiProvider

**Comment by jimmy1wu** [comment]
#libby

**Comment by jimmy1wu** [comment]
#libby

**Comment by jimmy1wu** [comment]
> The build jimmy1wu-20879-20220510-1647 https://wasrtc.hursley.ibm.com:9443/jazz/resource/itemOid/com.ibm.team.build.BuildResult/_HXZlUNCxEeyhmZ5zGPuLaA completed and has errors or failures.
> 
> For help analyzing your personal build, go to https://cognitive.hursley.ibm.com/buildAnalysis.html?uuid=_HXZlUNCxEeyhmZ5zGPuLaA

build failures are unrelated to these changes
